pkg/oauth/external/handler_test.go

```diff
@@ -149,3 +280,39 @@ func TestRedirectingStateError(t *testing.T) {
 		t.Errorf("Expected original error back, got %#v", err)
 	}
 }
+
+type provider struct {
+	name     string
+	username string
+	err      error
+	Provider
+}
+
+func (p provider) GetUserIdentity(*osincli.AccessData) (api.UserIdentityInfo, error) {
+	return &api.DefaultUserIdentityInfo{
+		ProviderName:     p.name,
+		ProviderUserName: p.username,
+	}, p.err
+}
+
+type userIdentityMapper struct{}
```

Comment:
This mocking here is the reason why we did not see that requests to the API would've been made in the error case, which would've been a good red flag.

Could you mock a real mapper with fakeclients instead?

Reply:
The code now traces calls to the mapper; the test cases check that the mapper is never called upon unsuccessful auth.